remove gsl-lite dependency

gridley opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

The C++ GSL does barely anything. Span is nice, but it's barely any effort to just make your own. Less dependencies is nice to have. I have had one issue requiring manual reconfiguration of GSL lite before when building openmc, so this will reduce the chance of others running into that.

Also, `assert` is arguably superior to `Expects` and `Ensures` provided by GSL. If we want, we can do `#define Expects assert`. But it's very nice to know many checks can be added with `assert` that vanish when compiling outside debug mode.

Lastly, since gsl::index is typedef'd to an int64_t 99.999999% of the time on modern computers, it replaced all instances with that. This would actually change behavior in certain cases on 32 bit computers, but using the 64 bit version can only help.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

gridley commented:
cpp-linter is really not great. It has false positives almost every time a change is made due to slight discrepancies in how different clang-format versions interpret things. I have gone through trying different clang-formats and still can't make it happy. It says to use version 15, but I can't find any straightforward way to install this on my mac.

I will have to return later to figure out what it's fussing about.

jtramm commented:
This is awesome -- thanks for removing this dependency!

When I run clang format with LLVM 15.0.0 or LLVM 18.1.0, I get the following diff:

```
diff --git a/src/mesh.cpp b/src/mesh.cpp
index 9ac60c864..2cb081bfd 100644
--- a/src/mesh.cpp
+++ b/src/mesh.cpp
@@ -3016,7 +3016,7 @@ void LibMesh::initialize()
     bin_to_elem_map_.reserve(m_->n_active_elem());
     elem_to_bin_map_.resize(m_->n_elem(), -1);
     for (auto it = m_->active_elements_begin(); it != m_->active_elements_end();
-      it++) {
+         it++) {
       auto elem = *it;
 
       bin_to_elem_map_.push_back(elem->id());
@@ -3175,7 +3175,7 @@ void LibMesh::set_score_data(const std::string& var_name,
   unsigned int std_dev_num = variable_map_.at(std_dev_name);
 
   for (auto it = m_->local_elements_begin(); it != m_->local_elements_end();
-    it++) {
+       it++) {
     if (!(*it)->active()) {
       continue;
     }
lines 1-22/22 (END)
```

Seems like LLVM 19 made some change to behavior here unfortunately.

jtramm commented:
I was able to push a fix for the formatting issue, but now it seems to be hitting a compile time issue when MPI is enabled. Could be an issue with some of the spans being changed to const types?

gridley commented:
Thanks for the fix John! I'll be able to fix that in a few days.

gridley commented:
I know I'm adding a PR right now on another little cleanup thing, but will come back to this soon to finish it.

gridley commented:
So, turns out that even though you would think that writing a source bank would be a const operation on the source sites themselves, that's actually not true if you're running in MPI mode since we do an `MPI_Recv` into that array before writing. So, I have added some documentation of this fact in `state_point.h`, and removed const where necessary.

This _should_ be good-to-go now. Sorry to take forever to come back to it!

gridley commented:
The CI fail here appears to be due to an upstream ncrystal change. I am happy to rebase if anyone finds a fix to this.

gridley commented:
I mentioned it in another PR, but it appears some upstream ncrystal change is the cause of the CI fail.

gridley commented:
I believe this is good to go now.

gridley commented:
Well, coveralls seems to have had an error this time. Guess I just need to trigger another test run and hope it works?

paulromano reviewed (approved):
Thanks @gridley!
